lldb/source/Plugins/TypeSystem/Swift/LRUCache.h

```diff
@@ -0,0 +1,136 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef liblldb_LRUCache_h_
+#define liblldb_LRUCache_h_
+
+#include "lldb/Utility/ConstString.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/STLFunctionalExtras.h"
+#include <list>
+#include <mutex>
+
+namespace lldb_private {
+namespace swift_demangle {
+
+/// A thread-safe Least Recently Used (LRU) cache implementation.
+///
+/// This class template implements an LRU cache with a fixed capacity, which
+/// supports insertion, retrieval, and a factory-based retrieval mechanism. It
+/// ensures thread safety through the use of a mutex lock. Keys are always
+/// `ConstString`
+template <typename Value>
+class LRUCache {
+public:
+  /// Constructs an instance with the specified capacity.
+  LRUCache(size_t capacity) : m_capacity(capacity) {}
+
+  /// Retrieves a value from the cache for the given key.
+  /// If the key is found in the cache, the corresponding value is returned and
+  /// the element is moved to the front of the LRU list, indicating it was
+  /// recently used. If the key is not found, std::nullopt is returned
+  std::optional<Value> Get(ConstString key) {
+    std::lock_guard lock{m_mutex};
+    auto map_it = m_map.find(key);
+    if (map_it == m_map.end())
+      return std::nullopt;
+    const auto &map_value = map_it->second;
+    MoveListElementToFront(map_value.list_it);
+    return map_value.value;
+  }
+
+  /// Inserts a key-value pair into the cache.
+  /// If the key already exists in the cache, its value is updated and the
+  /// element is moved to the front of the LRU list. If the key does not exist,
+  /// it is inserted into the cache. If the cache is at full capacity, the least
+  /// recently used element is evicted to make space for the new element.
+  void Put(ConstString key, const Value &value) {
+    if (m_capacity == 0)
+      return;
+    std::lock_guard lock{m_mutex};
+    auto map_it = m_map.find(key);
+    if (map_it != m_map.end()) {
+      auto &map_value = map_it->second;
+      map_value.value = value;
+      MoveListElementToFront(map_value.list_it);
+    } else {
+      InsertElement(key, value);
+    }
+  }
+
+  /// Retrieves a value from the cache or creates it if not present.
+  /// If the key is found in the cache, the corresponding value is returned and
+  /// the element is moved to the front of the LRU list. If the key is not
+  /// found, the given factory function is called to create a new value, which
+  /// is then inserted into the cache before being returned.
+  Value GetOrCreate(ConstString key, llvm::function_ref<Value()> factory) {
+    if (m_capacity == 0)
+      return factory();
+
+    {
+      std::lock_guard lock{m_mutex};
+      if (auto map_it = m_map.find(key); map_it != m_map.end()) {
+        const auto &map_value = map_it->second;
```

Comment:
Just thought about this: it'd be nice if in debug mode we assert that the cached value and the one created from the factory are the same (sorry for giving you more work!)

Reply:
Sure. The' Node' doesn't seem to provide an equality function. But it has `isSimilarTo` which in combination with a simple DFS should be enough to implement it in `SharedDemangledNode`.
I have a couple of questions then:
1) Do you want this check to be a part of the `LRUCache` class or in `GetCachedDemangledSymbol`? I'm asking because in `LRUCache` we generally don't need the `Value` type to be equatable, but with this check we will require that.
2) Under what flag should I add this check?
  - `!define(NDEBUG)` works in `ReleaseAssert` builds
  - `defined(LLDB_CONFIGURATION_DEBUG)` is defined only in debug builds (AFAICS)
  - Make a new macro like `LLDB_LRUCACHE_VALIDATE` that defaults to `NDEBUG` or `LLDB_CONFIGURATION_DEBUG`:
    ```
    #if !defined(LLDB_LRUCACHE_VALIDATE) && !defined(NDEBUG)
    #define LLDB_LRUCACHE_VALIDATE
    #endif
    ```
    This is more flexible but I doubt anyone will ever want to specify it separately.